### PR TITLE
Flatten version

### PR DESCRIPTION
**Pull request by calnix**
1. This is a generic implementation, since there was confusion between the Vault.sol and VaultV2.sol (stretch goals).
2. Used an FP (collateralLevel) instead of bufferNum/Denominator
3. Use of scalarFactor to rebase from priceFeed multiplication/division operations. 
5. Addressed most other comments

### CODE REVIEW

**Comment by calnix** [comment]
For testCannotBorrowInExcess(), it is interesting that this does not work as a fuzzing implementation:
> vm.assume(excessDebt > maxDebt && excessDebt < type(uint).max);	

Results in an (Arithmetic over/underflow). Similar with testCannotBorrowExceedingMargin.

Strange, as I did not face this in my initial implementation - not too sure why it would be different now

**Comment by calnix** [comment]
Added tests suggested by Richie.

@Alberto: I moved getMaxDebt() into my Test file as it was very helpful with some of the scenario tests.
Made me rethink its necessity(or lack of) in the Vault contract itself.

Would it not be useful for users to be able to check on their max possible borrow-able amount given current market conditions?

File: src/Vault.sol [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        deposits[msg.sender] += collateralAmount;
+        bool sent = collateral.transferFrom(msg.sender, address(this), collateralAmount);
+        require(sent, "Deposit failed!");  
+        emit Deposit(address(collateral), msg.sender, collateralAmount);
+    }
+    
+    ///@notice Users borrow debt asset calculated based on collateralization level and their deposits 
+    ///@dev See getMaxDebt() for colleteralization calculation
+    ///@param debtAmount Amount of debt asset to borrow
+    function borrow(uint debtAmount) external {
```
Comment (alcueca) [inline src/Vault.sol]: The user can borrow several times using the same collateral, the collateralisation check should happen on `debts[msg.sender] + debtAmount`

File: src/Vault.sol [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+    ///@notice Price is returned as an integer extending over its decimal places
+    ///@dev Calculates collateral required to support existing debt position at current market prices 
+    ///@param debtAmount Amount of debt asset to support
+    /*Note: Example calculation using collateralLevel (e.g. DAI/WETH):
+    *       baseCollateralRequired = 1 WETH (1e18), collateralLevel = 80% (8e17), scalarFactor = 10**18  (DAI/WETH is 18dp)
+    *       1 WETH (1e18) * 80% (8e17) / 10**18 = 8e(18+17) / 10**18 = 8e17  => 80% of 1 WETH
+    *       Since we want to have a buffer, we will DIVIDE by collateralLevel: 1/0.8 = 1.25 (20% buffer, 80% collateralization level)
+    */
+    function getCollateralRequired(uint debtAmount) public view returns(uint) {
```
Comment (alcueca) [inline src/Vault.sol]: Maybe changing this function for something like `function isCollateralized(address user) view returns (bool)` or `isCollateralized(uint256 debt, uint256 collateral) view returns (bool)` might make things simpler.

File: src/Vault.sol [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        (,int price,,,) = priceFeed.latestRoundData();
+        uint baseCollateralRequired = debtAmount * uint(price) / scalarFactor;
+        uint bufferCollateralRequired = (baseCollateralRequired * scalarFactor) / collateralLevel; 
+        return bufferCollateralRequired = scaleDecimals(bufferCollateralRequired, collateral.decimals(), debt.decimals());
```
Comment (alcueca) [inline src/Vault.sol]: Each one of these `.decimals()` is a CALL to an external contract, and expensive at 2100+ gas. You should read the decimals in the constructor and store them, since they won't change.

File: src/Vault.sol [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        }
+    }
+
+   
+    ///@notice Rebasement is necessary when utilising assets with divergering decimal precision
+    ///@dev For rebasement of the trailing zeros which are representative of decimal precision
+    function scaleDecimals(uint integer, uint from, uint to) public pure returns(uint) {
+        //downscaling | (to-from) => (-ve)
+        if (from > to ){ 
+            return integer * 10**(to-from);
```
Comment (alcueca) [inline src/Vault.sol]:
```suggestion
            return integer * 10**(to - from);
```

File: src/Vault.sol [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    ///@param liquidatedCollateralAmount The amount of collateral received by the liquidator
+    event Liquidation(address indexed collateralAsset, address indexed debtAsset, address indexed user, uint debtToCover, uint liquidatedCollateralAmount);
+    
+    ///@notice To allow for margined collateralization; similar to FX brokers
+    ///@dev Fixed-point number; to set collateralization level of debt engine
+    /*Note: collateralLevel is a fixed-point number representation of a decimal (e.g. 0.8), | collateralLevel = range(0, 1e18]
+    *       therefore value supplied has to be order of magnitude smaller than the precision of collateral asset itself.
+    *       Example: 1 DAI (1e18) | 80%: 0.8 DAI (8e17) -> therefore, collateralLevel = 8e17
+    */     
+    uint public collateralLevel;
```
Comment (alcueca) [inline src/Vault.sol]: On blockchain it is much more common for the collateralisation ratio to be above 100%. I don't think anyone has managed to build a protocol offering undercollateralised loans yet.

Comment (calnix) [reply src/Vault.sol]: Really? Shouldn't this be an easier proposition - also safer, from both a product offering as well as risk

File: src/Vault.sol (hunk 2 above)
Comment (calnix) [reply src/Vault.sol]: Wouldn't it be necessary for the end-user to know how much collateral would be required for a certain debt amount, before committing to said debt amount? Like a preview assessment of sorts before pulling the trigger.

File: src/Vault.sol (hunk 1 above)
Comment (calnix) [reply src/Vault.sol:94]: so that's why i had `getMaxDebt()` to check the level of available collateral given deposits and existing debt.
it would serve as a check in borrow

Comment (calnix) [reply src/Vault.sol]: You're right, good catch. After removing `maxDebt` I did not transform `borrow()` accordingly to work without it.

**Comment by calnix** [comment]
I'm stuck on this single test function testBorrowLessAtNewRate(), and can't figure out whats going wrong.

For some reason, this fails: 

```
        uint currentAvailableDebt = StateRateChange2.getMaxDebt2(user); 
        uint collateralRequired = vault.getCollateralRequired(currentAvailableDebt);
        uint availableCollateral = vault.deposits(user) - vault.getCollateralRequired(vault.debts(user));
        // assertEq(currentAvailableDebt, availableCollateral);
```

But i'm able to borrow some value up to availableCollateral.

File: src/Vault.sol (hunk 5 above)
Comment (calnix) [reply src/Vault.sol]: Dropped the collateralization ratio aspect.

File: src/Vault.sol (hunk 1 above)
Comment (alcueca) [reply src/Vault.sol]: Maybe I'm turning into a better auditor :D

File: src/Vault.sol (hunk 2 above)
Comment (alcueca) [reply src/Vault.sol]: That's a good point, and is something that we did for ERC4626 with all those `max*` functions.

The answer to your question is "it depends". If your end user is going to use a frontend, that frontend can probably calculate those numbers.

Less code in a smart contract is generally a good thing, but when we are talking about view functions that have little risk.

This kind of helper view functions become useful with contract integration.  We implemented [some](https://github.com/yieldprotocol/vault-v2/blob/0b0926c3f7c0f06677a450a52a36352334b71cc5/contracts/Cauldron.sol#L271) in our Collateralized Debt Engine, and they ended up being quite useful.

Long story short, let's simplify the code logic as much as we can (one single function to judge collateralization) but be quite open to introduce zero-risk helper functions that give information that might be useful to integrators.

Comment (alcueca) [reply src/Vault.sol]: So I guess that you should do a `isCollateralized` function for internal use, just because of how much it simplifies your core logic, and then you can implement `minCollateral` and `maxDebt` external functions, because they will be useful.

Comment (calnix) [reply src/Vault.sol]: I'm unsure how I could be further simplifying `getCollateralRequired()` into `isCollateralized` as suggested.
I come back to the same functionality out of necessity, given that `borrow()` relies on `getCollateralRequired()` as a collateralization check.  

Else I would run into the earlier issue of user being able to repeatedly borrow using the same collateral.

However, i could lean out `borrow()` like so:

```
    function borrow(uint debtAmount) external {      
        
        require(getCollateralRequired(debtAmount + debts[msg.sender]) <= deposits[msg.sender], "Insufficient collateral!");

        debts[msg.sender] += debtAmount;
        bool sent = debt.transfer(msg.sender, debtAmount);
        require(sent, "Borrow failed!");       
        emit Borrow(address(debt), msg.sender, debtAmount);
    }
```

```
    function getCollateralRequired(uint debtAmount) public view returns(uint) {
        (,int price,,,) = priceFeed.latestRoundData();
        uint collateralRequired = debtAmount * uint(price) / scalarFactor;
        return collateralRequired = scaleDecimals(collateralRequired, collateralDecimals, debtDecimals);
    }
```

Would this be acceptable?
